api/src/main/java/edu/wpi/first/shuffleboard/api/data/DataTypes.java

```diff
@@ -1,95 +1,225 @@
+    }
+    return defaultInstance;
+  }
+
+  /**
+   * Creates a new data type registry. The registry will initially contain {@link #None}, {@link #All},
+   * {@link #Unknown}, and {@link #Map}, none of why may be unregistered.
+   */
+  public DataTypes() {
     register(None);
+    register(All);
     register(Unknown);
-    register(Number);
-    register(String);
-    register(Boolean);
-    register(NumberArray);
-    register(StringArray);
-    register(BooleanArray);
     register(Map);
-    register(SendableChooser);
-    register(AnalogInput);
-    register(SpeedController);
-    register(Encoder);
   }
 
-  private DataTypes() {
+  /**
+   * Sets the default instance to use. <strong>This may only be called from tests</strong>.
+   *
+   * @throws IllegalStateException if not called from a test
+   */
+  @VisibleForTesting
+  public static void setDefault(DataTypes instance) {
+    TestUtils.assertRunningFromTest();
+    defaultInstance = instance;
   }
 
   /**
    * Registers the given data type.
    *
    * @param dataType the data type to register
    */
-  public static void register(DataType<?> dataType) {
+  @Override
+  public void register(DataType dataType) {
+    requireNonNull(dataType, "dataType");
+    if (isRegistered(dataType)) {
+      throw new IllegalArgumentException("Data type " + dataType + " has already been registered");
+    }
     dataTypes.put(dataType.getName(), dataType);
+    typeCache.put(dataType.getJavaClass(), Optional.of(dataType));
+    addItem(dataType);
+  }
+
+  @Override
+  public void unregister(DataType dataType) {
+    dataTypes.remove(dataType.getName());
+    typeCache.remove(dataType.getJavaClass());
+    removeItem(dataType);
   }
 
   /**
    * Gets the data type with the given name.
    */
-  public static Optional<DataType<?>> forName(String name) {
+  public Optional<DataType> forName(String name) {
     return Optional.ofNullable(dataTypes.get(name));
   }
 
   /**
+   * Gets the data type most relevant to a Java class.
+   */
+  @SuppressWarnings("unchecked")
+  public <T> Optional<DataType<T>> forJavaType(Class<T> type) {
+    if (type.isPrimitive()) {
+      return forJavaType((Class) Primitives.wrap(type));
+    }
+    return (Optional) typeCache.computeIfAbsent(type, __ -> {
+      if (DataType.class.isAssignableFrom(type)) {
+        return forType((Class<DataType>) type);
+      }
+      Optional<DataType> naive = dataTypes.values().stream()
+          .filter(t -> type.equals(t.getJavaClass()))
+          .findAny();
+      if (naive.isPresent()) {
+        return naive;
+      } else {
+        // Check Java class hierarchy
+        Comparator<Class<?>> classComparator = closestTo(type);
+        List<DataType> sorted = dataTypes.values().stream()
+            .filter(t -> t != All)
+            .sorted((t1, t2) -> classComparator.reversed().compare(t1.getJavaClass(), t2.getJavaClass()))
+            .collect(Collectors.toList());
+        return Optional.of(sorted.get(0));
```

Comment:
Ah, that may work. `ofNullable` is pointless since elements should never be null